cpp/src/io/json/reader_impl.cu

```diff
@@ -0,0 +1,552 @@
+
+    rmm::device_vector<cudf::experimental::io::json::ColumnInfo> d_column_infos(num_columns, cudf::experimental::io::json::ColumnInfo{});
+    cudf::experimental::io::json::gpu::detect_data_types(d_column_infos.data().get(),
+                         static_cast<const char*>(data_.data()), data_.size(), 
+                         opts_, num_columns,
+                         rec_starts_.data().get(), rec_starts_.size(), stream);
+    thrust::host_vector<cudf::experimental::io::json::ColumnInfo> h_column_infos = d_column_infos;    
+
+    for (const auto &cinfo : h_column_infos) {
+      if (cinfo.null_count == static_cast<int>(rec_starts_.size())) {
+        // Entire column is NULL; allocate the smallest amount of memory
+        dtypes_.push_back(data_type(INT8));
+      } else if (cinfo.string_count > 0) {
+        dtypes_.push_back(data_type(STRING));
+      } else if (cinfo.datetime_count > 0) {
+        dtypes_.push_back(data_type(TIMESTAMP_MILLISECONDS));
+      } else if (cinfo.float_count > 0 || (cinfo.int_count > 0 && cinfo.null_count > 0)) {
+        dtypes_.push_back(data_type(FLOAT64));
+      } else if (cinfo.int_count > 0) {
+        dtypes_.push_back(data_type(INT64));
+      } else if (cinfo.bool_count > 0) {
+        dtypes_.push_back(data_type(BOOL8));
+      } else {
+        CUDF_FAIL("Data type detection failed.\n");
+      }
+    }
+  }  
+}
+
+/**
+ * @brief Parse the input data and store results a table
+ *       
+ * @param[in] stream Cuda stream to run kernels on
+ *
+ * @return table_with_metadata struct
+ **/
+table_with_metadata reader::impl::convert_data_to_table(cudaStream_t stream) {  
+  const auto num_columns = dtypes_.size();
+  const auto num_records = rec_starts_.size();
+   
+  // alloc output buffers. 
+  std::vector<column_buffer> out_buffers;
+  for (size_t col = 0; col < num_columns; ++col) {
+    out_buffers.emplace_back(dtypes_[col], num_records, stream, mr_);
+  }
+
+  thrust::host_vector<data_type> h_dtypes(num_columns);
+  thrust::host_vector<void *> h_data(num_columns);
+  thrust::host_vector<bitmask_type *> h_valid(num_columns);
+
+  for (size_t i = 0; i < num_columns; ++i) {
+    h_dtypes[i] = dtypes_[i];
+    h_data[i] = out_buffers[i].data();
+    h_valid[i] = out_buffers[i].null_mask();
+  }
+
+  rmm::device_vector<data_type> d_dtypes = h_dtypes;
+  rmm::device_vector<void *> d_data = h_data;
+  rmm::device_vector<cudf::bitmask_type *> d_valid = h_valid;
+  rmm::device_vector<cudf::size_type> d_valid_counts(num_columns, 0);
+
+  cudf::experimental::io::json::gpu::convert_json_to_columns( data_, d_dtypes.data().get(), d_data.data().get(),
+                        num_records, num_columns, 
+                        rec_starts_.data().get(),
+                        d_valid.data().get(), d_valid_counts.data().get(), 
+                        opts_, stream);  
+  CUDA_TRY(cudaDeviceSynchronize());
```

Comment:
This should probably be using cudaStreamSynchronize(stream) ?

Reply:
Yep.  Fixed.